tensorflow/core/kernels/segment_reduction_ops.h

```diff
@@ -76,6 +79,7 @@ struct AtomicSumOpGpu {
                                                         const T& value) {
     GpuAtomicAdd(dest, value);
   }
+  const bool deterministic_for_float = false;
```

Comment:
This should say `is_associative` since individual floating point additions are deterministic.  And I'd suggest writing this as:

```
constexpr bool is_associative = std::is_floating_point<T>::value;
```

so that the "caller" does not have to check `T` again.

Same for the other instances.

Reply:
Very nice. Will do.

Reply:
I made this into a `static constexpr` (class parameter). `constexpr` is not allowed on a non-static member of a class presumably because it has to be assigned (even if const) when the class is instantiated (at runtime and not during compilation). This change, which I think is what we were looking for, may facilitate more compile-time elimination of of some of the logic used to decide branching.

Reply:
Also, I used `std::is_integral<T>::value` because `!std::is_floating_point<T>::value` will not include the complex types. I'm also more confident that all of the integral types are associative than I am confident that all types other than the floating point types (including the complex types) are associative.